### DISCUSSION TITLE
Simulation mode questions

### DISCUSSION THREAD

**Discussion by guidopowell** [asreview/asreview#1802] [Q&A]
Hi I have a few questions about simulation mode.

1) From what I understand, demonstrations and tutorials on simulation mode make use of the relevance label of each abstract based on the final inclusion of the abstract (after the full text is read), not the label assigned during the abstract screening stage. 

This will not be an accurate representation of what the tool will do in Oracle mode since in the real world, active learning results are determined by relevance at the abstract level.

I can get around this by running a simulation with the abstract relevance labels, and do my own evaluation with the relevance labels from the full text stage. But I am just surprised this hasn't been discussed anywhere. I am sure it must have an impact on the performance of the model.

2) I also have a question about whether simulation mode stops at any point or whether the active learning continues iterating until the last abstract. 

Thank you!

**Comment by jteijema** [comment]
Hi @guidopowell,

Thanks for asking!

1. Some of our testing datasets are indeed title abstract (tiab) inclusions and not full text, for this reason. But not all of them. We've considered this, and yes, this will introduce some noise into the process. Noise in the form of an abstract that was seemingly relevant but labeled irrelevant because information that was not in the abstract. Luckily the models we're using are quite resistant to _some_ levels of noise, otherwise we'd have a hard time with real world applications hahaha.
2. Active learning will keep going until the last paper! However, we've been looking into something we're calling the rank order stability, i.e. how much the ranking changes with each iteration. So far we've seen that it stabilized after a while (naturally), but no definitive answers have been found yet.

**Comment by guidopowell** [comment]
Thank you @jteijema for this helpful answer. 

1) I think that beyond providing labels of title/abstract inclusion, the label of full text inclusion would have to provided as well for a proper evaluation. We need to see how the model behaves when active learning is iteratively fed relevance labels based on title/abstract (how it is used in the real world) and subsequently to evaluate this simulation against the full text labels, given we don't actually care about recall of title/abstract inclusions.  

We have many datasets where there are 5-6 times as many title/abstract inclusions vs full text inclusions. If the active learning is being fed these additional labels there might be more than a bit of noise (especially for WSS) compared to active learning based on the few full text labels (the 5-fold increase in positives should incur a large number of false positives as well). 

A conservative but not unrealistic perspective is that none of the simulation results are appropriate estimates of the model performance. I would really love to contribute with analyses to better estimate the performance. Unlike what I said in my initial post, it turns out that it would not be very simple to do as the model output is nested somewhere in the .asreview file. It's not clear yet to me how to extract the appropriate results. 

2) Thank you, looking forward to learning the results of this development. 

Appreciate the discussion and continued involvement in this software.

**Reply by guidopowell** [reply]
As a follow-up to this. I imagine what I would need is to extract from the .asreview file the results of the simulation based off the title/abstract labels, and compare those to my own full text labels (maybe joining the datasets externally, hopefully based on some identifier). 


Now, based on your answer to question 2, I am not sure it's clear what output of the simulation I would need. 
In the real world, people will stop the active learning after some stopping rule (e.g., number of irrelevant in a row). But the output of a simulation that is run until the last article will score differently from one that is stopped somewhere. And I can imagine it is not possible to record and share the ranking at each iteration. I guess this is definitely something you have thought through. 

Just in case it isn't clear here is a scenario. If there are let's say 5 relevant articles hidden far away in the ranking, and they don't get brought up until say 65% of screening is completed, a real world user may never identify them if prior to that 65% mark the active learning presents them with 50 irrelevant articles and that is their stopping rule. The simulation output shows us the results of the most sensitive possible stopping rules, not those that people may use. 

I can think of some ways in which this question interacts with the one about title/abstract labels, but I think it's something I would hope to understand independent of that one first.

Thank you!

**Reply by jteijema** [reply]
[_Bron, M. P., van der Heijden, P. G., Feelders, A. J., & Siebes, A. P. (2024). Using Chao's Estimator as a Stopping Criterion for Technology-Assisted Review. arXiv preprint arXiv:2404.01176._ ](https://arxiv.org/abs/2404.01176)<- a beautiful paper about the impact of different stopping rules on the recall. Exactly as you say, stopping rules are bound to artificially inflate performance for real live applications because unfound papers are deemed irrelevant and therefore excluded from the performance statistics. So, sensitivity of the stopping rule does indeed have impact on performance. Tha paper also has some of the best graphics I've seen in a long time.

**Reply by jteijema** [reply]
1. Well, lets start with a dataset that has both labels available, and visualize that. Are the full-text relevant document found in a significant pattern when compared to the TIAB inclusions? I think that's a good start for this question. 

I agree that's its a bit more than a little noise, but the papers DO overlap significantly of course. And there's another nuance to include;

The classification of ASReview is not one optimized for relevance predictions. We optimize for the best order in which to show the user all papers they have to screen, which can in fact be quite far removed from the perfect predictions of relevance.

**Reply by jteijema** [reply]
Getting the ranking every iteration is possible. I can't imagine it being a coincidence at this point, but the thread before yours (#1801)  asked this very same question. We've answered there, with some code as to how to store the ranking at each step.

**Reply by jteijema** [reply]
And finally, the asreview file is quite interpretable using our API. Opening the state (https://asreview.readthedocs.io/en/latest/generated/asreview.open_state.html) and using the state (https://asreview.readthedocs.io/en/latest/generated/asreview.state.SQLiteState.html) might give some great insight, together with using our metrics package (https://github.com/asreview/asreview-insights) would surely be a good start.

**Reply by guidopowell** [reply]
Thank you @jteijema for the responses.

That paper on stopping rules is quite impressive I will have to look at it carefully. 

And I believe it is a coincidence I asked the same question as the previous one , I don't think I had seen it!

I will have to spend some more time evaluating simulations from the links on the API. Thank you for sharing!

**Reply by jteijema** [reply]
Perfect! Thank you for your interest. And if you find anything, please let us know, we're always interested in updates from our users/researchers. And if you're interested, [Teijema, J. J., de Bruin, J., Bagheri, A., & van de Schoot, R. (2023, November 1). Large-Scale Simulation Study of Active Learning models for Systematic Reviews. https://doi.org/10.31234/osf.io/2w3rm](https://osf.io/2w3rm) has simulation results for our simulations on synergy.
